### PR TITLE
feat(kafka sink): Add support for SASL

### PR DESCRIPTION
**Pull request by 0x003e**
Hello,

https://github.com/timberio/vector/issues/1277

Some people need SASL Authentification on Kafka:

  * feat(kafka sink): Add support for SASL 
  * docs: https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
   >
security.protocol | * | plaintext, ssl, **sasl_plaintext**, **sasl_ssl** | plaintext | high | Protocol used to communicate with brokers. Type: enum value
-- | -- | -- | -- | -- | --

https://github.com/0x003e/vector/commit/5cc42f085fa02647f2d2569ef38ed4a6e7be9228#diff-3380bc35763ec3d083d279f313dfa4f4R135
```rust
security_protocol = if tls_enabled && sasl_enabled { "sasl_ssl" }
        else if tls_enabled && !sasl_enabled { "ssl" }
        else if !tls_enabled && sasl_enabled { "sasl_plaintext" }
        else { "plaintext" };
        client_config.set("security.protocol", security_protocol);
```

### CODE REVIEW

File: website/docs/reference/sinks/kafka.md [hunk 1]
```diff
@@ -120,6 +120,21 @@ import CodeHeader from '@site/src/components/CodeHeader';
...
+    enabled = true # default
+
+    # OPTIONAL - Username
+    username = "username" # example, no default
+
+    # OPTIONAL - Password
+    password = "password" # example, no default
+    
+    # OPTIONAL - mechanism
+    mechanism = "scram-sha-256" # example, no default
```
Comment (lukesteensen) [inline website/docs/reference/sinks/kafka.md]: This file is actually generated automatically from the data in `.meta/sinks/kafka.toml`, so if you add the new options there they will automatically get propagated where they need to go.

You can see an example of a config sub-group [here](https://github.com/timberio/vector/blob/e76083548a2d46664acd67a8e40f1835614d94c5/.meta/sinks/elasticsearch.toml#L47-L62).

Comment (0x003e) [reply website/docs/reference/sinks/kafka.md]: Hi,
Actually librdkafka return a error if mechanism not corrrect when healthcheck = true

About the mistake in documentation : I'll correct this commit with a 'amend' command this week-end

**Comment by binarylogic** [comment]
Thanks for this @0x003e ! Do you  plan on making the remaining changes? If not, we can assign someone to get this completed, but would love your help if you're willing.

File: website/docs/reference/sinks/kafka.md (hunk 1 above)
Comment (0x003e) [reply website/docs/reference/sinks/kafka.md]: If mechanism are invalid (only UPERCASE):

image

**Comment by 0x003e** [comment]
Hi @binarylogic, @lukesteensen 

I have made the changes

File: src/sinks/kafka.rs [hunk 2]
```diff
@@ -337,6 +365,7 @@ mod integration_test {
             encoding: Encoding::Text,
             key_field: None,
             tls,
+            sasl: None,
```
Comment (Hoverbear) [inline src/sinks/kafka.rs]: Would it be practical to test sasl on in a test?

**Comment by Hoverbear** [comment]
I resolved merge conflicts on this. :)

**Comment by Hoverbear** [comment]
I am frankly not sure what's happening with the CI on this.

**Comment by Hoverbear** [comment]
Looks like this PR will also need to pull in `sasl2` as a dependency, probably with `cryrus-sasl`.

**Comment by binarylogic** [comment]
@0x003e we're closing this since we haven't heard from you. It's also drifted quite a bit from master. Finally, we have a new [`librdkafka_options` option](https://vector.dev/docs/reference/sources/kafka/#librdkafka_options) that you can use to pass `librdkafka` flags.

**Comment by 0x003e** [comment]
Thank you for the new **librdkafka_options**, and I am sorry for the delay. 

Howerver, can you add '**sasl**' features for rdkafka in the Cargo.toml ? :

rdkafka = { version = "0.23.1", features = ["libz", "ssl", "zstd", **'sasl'** ], optional = true }

**Comment by lukesteensen** [comment]
@Hoverbear last time I tried adding the `sasl` feature to `rdkafka` I ran into some build issues due to requiring an additional C dependency. Does this seem like something that'd be easier with our new fancy build tooling?

**Comment by Hoverbear** [comment]
@lukesteensen , can we defer it to @fanatid ? :) He's volunteered and I'll help!

**Comment by lukesteensen** [comment]
Sounds good to me!

**Pull request closed without merging**